src/main/java/io/confluent/idesidecar/restapi/models/graph/RealDirectFetcher.java

```diff
@@ -106,64 +106,72 @@ public List<DirectConnection> getConnections() {
+                  kafkaConfig.bootstrapServers(),
+                  connectionId
+              )
+            ).map(cluster -> {
+              // Emit an event that this cluster was loaded
+              onLoad(connectionId, cluster);
+              // And return the cluster
+              return cluster;
+            });
+      }).orElseGet(() -> Uni.createFrom().nullItem());
     }
+    // Unexpectedly not a direct connection
+    Log.errorf("Connection with ID=%s is not a direct connection.", connectionId);
     return Uni.createFrom().nullItem();
   }
 
   public Uni<DirectSchemaRegistry> getSchemaRegistry(String connectionId) {
-    var spec = connections.getConnectionSpec(connectionId);
-    DirectSchemaRegistry cluster = null;
-    if (spec != null && DIRECT.equals(spec.type()) && spec.schemaRegistryConfig() != null) {
-      var srConfig = spec.schemaRegistryConfig();
-      var ccloudEndpoint = srConfig.asCCloudEndpoint();
-      if (ccloudEndpoint.isPresent()) {
-        var endpoint = ccloudEndpoint.get();
-        cluster = new DirectSchemaRegistry(
-            endpoint.clusterId().toString(),
-            endpoint.getUri().toString(),
-            connectionId
-        );
-      } else {
-        cluster = new DirectSchemaRegistry(
-            orDefault(srConfig.id(), () -> connectionId + "-schema-registry"),
-            srConfig.uri(),
-            connectionId
-        );
-      }
-    }
-    if (cluster != null) {
-      onLoad(connectionId, cluster);
-      return Uni.createFrom().item(cluster);
+    var state = connections.getConnectionState(connectionId);
+    if (state instanceof DirectConnectionState directState) {
+      var spec = state.getSpec();
+      // Use the admin API to obtain the cluster ID
+      return directState.withSchemaRegistryClient(srClient -> {
+        Uni<DirectSchemaRegistry> result;
+        try {
+          var mode = srClient.getMode(); // unused, but we know we can connect
+          var srConfig = spec.schemaRegistryConfig();
+          var ccloudEndpoint = srConfig.asCCloudEndpoint();
+          DirectSchemaRegistry cluster;
+          if (ccloudEndpoint.isPresent()) {
+            var endpoint = ccloudEndpoint.get();
+            cluster = new DirectSchemaRegistry(
+                endpoint.clusterId().toString(),
+                endpoint.getUri().toString(),
+                connectionId
+            );
+          } else {
+            cluster = new DirectSchemaRegistry(
+                orDefault(srConfig.id(), () -> connectionId + "-schema-registry"),
+                srConfig.uri(),
+                connectionId
+            );
+          }
+          result = Uni.createFrom().item(cluster);
+        } catch (Throwable t) {
+          Log.debugf("Failed to create schema registry client for connection ID=%s", connectionId, t);
+          result = Uni.createFrom().nullItem();
+        }
+        return result;
+      }).orElseGet(() -> Uni.createFrom().nullItem());
```

Comment:
Perhaps the "failed to create client" debug log should go in this supplier function.

```suggestion
      }).orElseGet(() -> {
        Log.debugf("Failed to create schema registry client for connection ID=%s", connectionId, t);
        Uni.createFrom().nullItem();
      });
```

Reply:
Any _uncaught_ exception is logged as an error in the `withSchemaRegistry(...)` method, and the contract is that the function should handle all exceptions on its own.

What's interesting is that we can create instances of the `KafkaAdminClient` or the `SchemaRegistryClient` with incorrect credentials or wrong information (e.g., bootstrap servers, URL), but only when we try and use those clients will they actually try to connect by making calls over the network.

This is why the `withSchemaRegistry(...)` and `withAdminClient(...)` methods in `DirectConnectionState` create the client objects safely before calling the supplied function, **and** expect the supplied function to handle all exceptions. This means that any exception caught in the `with...` methods will be only from the supplied function, meaning it is unexpected and therefore it's a bug that needs to be fixed. And that's why such exceptions are logged as errors.

So, the `orElseGet(...)` block here is called only when there is no schema registry configuration for this connection, or if there is an uncaught exception in the function (at which point it's only logged). I can add comments and log debug messages to help make that clear, but the log message should _not_ be "failed to create a connection" in the `orElseGet(...)` block.

Reply:
Actually, my statement above about always being able to create a KafkaAdminClient is not quite correct: credentials are indeed only checked once you do something with the client, but the bootstrap server addresses are resolved immediately upon creation of the KafkaAdminClient, so I'll handle that in this PR.

Reply:
Improved how exceptions are handled, making it much easier to handle all errors in one spot.